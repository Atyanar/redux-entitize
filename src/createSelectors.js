// @flow

import { denormalize } from "normalizr";

import { type StateType, type SchemaEntitiesMapType } from "./reducer";
import type { EntityType, SchemaMapType } from "./types";

export default function createSelectors(schemaMap: SchemaMapType) {
  function selectEntity(
    state: { entities: StateType },
    schema: string,
    id: string
  ): ?EntityType {
    ensureExistingSchema(state, schema);

    return (
      denormalize(
        { [schema]: [id] },
        { [schema]: [schemaMap[schema]] },
        state.entities
      )[schema][0] || null
    );
  }

  function selectEntities(
    state: { entities: StateType },
    schema: string,
    ids?: string[]
  ): EntityType[] {
    const schemaEntities = selectSchemaEntitiesMap(state, schema);
    const relevantIDs = ids || Object.keys(schemaEntities);

    return denormalize(
      { [schema]: relevantIDs },
      { [schema]: [schemaMap[schema]] },
      state.entities
    )[schema];
  }

  function selectSchemaEntitiesMap(
    state: { entities: StateType },
    schema: string
  ): SchemaEntitiesMapType {
    const schemaEntities = state.entities[schema];

    ensureExistingSchema(state, schema);

    return schemaEntities;
  }

  function ensureExistingSchema(
    state: { entities: StateType },
    schema: string
  ) {
    if (!state.entities[schema]) {
      throw new Error(
        `Schema '${schema}' is unkown. Schemas in state are: [${Object.keys(
          state.entities
        ).join(", ")}]`
      );
    }
  }

  return {
    selectEntity,
    selectEntities
  };
}
