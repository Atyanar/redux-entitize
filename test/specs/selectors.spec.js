// @flow

import { schema } from "normalizr";

import createSelectors from "../../src/createSelectors";

describe("selectors", () => {
  describe("when initialized with an array entity schema", () => {
    function setup() {
      const schemaMap = {
        articles: new schema.Entity("articles", {
          comments: [new schema.Entity("comments")]
        })
      };
      const state = {
        entities: {
          articles: {
            article_1: {
              id: "article_1",
              title: "Foo Bar 1",
              comments: ["comment_1", "comment_2"]
            },
            article_2: {
              id: "article_2",
              title: "Foo Bar 2",
              comments: []
            },
            article_3: {
              id: "article_3",
              title: "Foo Bar 3",
              comments: []
            }
          },
          comments: {
            comment_1: {
              id: "comment_1",
              title: "Good"
            },
            comment_2: {
              id: "comment_2",
              title: "Bad"
            }
          }
        }
      };

      return {
        state,
        ...createSelectors(schemaMap)
      };
    }
    describe("selectEntity", () => {
      describe("when called with non-existent schema", () => {
        test("throws an error", () => {
          const { selectEntity, state } = setup();

          expect(() =>
            selectEntity(state, "unkownSchema", "irrelevant_id")
          ).toThrowErrorMatchingSnapshot();
        });
      });

      describe("when called with existent id", () => {
        test("returns the entity data with nested entities", () => {
          const { selectEntity, state } = setup();

          expect(
            selectEntity(state, "articles", "article_1")
          ).toMatchSnapshot();
        });
      });

      describe("when called with non-existent id", () => {
        test("returns null", () => {
          const { selectEntity, state } = setup();

          expect(
            selectEntity(state, "articles", "article_non_existent")
          ).toEqual(null);
        });
      });
    });
    describe("selectEntities", () => {
      describe("when called with only the schema", () => {
        test("returns array of all entities' data with nested entities", () => {
          const { selectEntities, state } = setup();

          expect(selectEntities(state, "articles")).toMatchSnapshot();
        });
      });
      describe("when called with the schema and an array of IDs", () => {
        test("returns array of selected and nested entities' data", () => {
          const { selectEntities, state } = setup();

          expect(
            selectEntities(state, "articles", ["article_2", "article_3"])
          ).toMatchSnapshot();
        });
      });
    });
  });
});
