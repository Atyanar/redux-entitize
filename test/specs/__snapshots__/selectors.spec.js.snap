// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectors when initialized with an array entity schema selectEntities when called with only the schema returns array of all entities' data with nested entities 1`] = `
Array [
  Object {
    "comments": Array [
      Object {
        "id": "comment_1",
        "title": "Good",
      },
      Object {
        "id": "comment_2",
        "title": "Bad",
      },
    ],
    "id": "article_1",
    "title": "Foo Bar 1",
  },
  Object {
    "comments": Array [],
    "id": "article_2",
    "title": "Foo Bar 2",
  },
  Object {
    "comments": Array [],
    "id": "article_3",
    "title": "Foo Bar 3",
  },
]
`;

exports[`selectors when initialized with an array entity schema selectEntities when called with the schema and an array of IDs returns array of selected and nested entities' data 1`] = `
Array [
  Object {
    "comments": Array [],
    "id": "article_2",
    "title": "Foo Bar 2",
  },
  Object {
    "comments": Array [],
    "id": "article_3",
    "title": "Foo Bar 3",
  },
]
`;

exports[`selectors when initialized with an array entity schema selectEntity when called with existent id returns the entity data with nested entities 1`] = `
Object {
  "comments": Array [
    Object {
      "id": "comment_1",
      "title": "Good",
    },
    Object {
      "id": "comment_2",
      "title": "Bad",
    },
  ],
  "id": "article_1",
  "title": "Foo Bar 1",
}
`;

exports[`selectors when initialized with an array entity schema selectEntity when called with non-existent schema throws an error 1`] = `"Schema 'unkownSchema' is unkown. Schemas in state are: [articles, comments]"`;
